torch/_dynamo/compile_package.py

```diff
@@ -0,0 +1,756 @@
+        assert isinstance(meta_tensor, torch.Tensor)
+        self.builder.TENSOR_MATCH(
+            guard, value=torch.empty_like(meta_tensor, device=device)
+        )
+
+    def TYPE_MATCH(self, guard, metadata):
+        from torch._dynamo.guards import get_verbose_code_parts
+        from torch._dynamo.source import AttrSource
+        from torch._guards import Guard
+
+        module, qualname = metadata
+        cls_source = AttrSource(guard.originating_source, "__class__")
+
+        # TODO Should we just create a new type of FQN-matching guard for this?
+        def add_guard(attr, val):
+            source = AttrSource(cls_source, attr)
+            guard = Guard(source, None)
+            ref = self.builder.arg_ref(guard)
+            code = [f"{ref} == {val!r}"]
+            self.builder.get_guard_manager(guard).add_equals_match_guard(
+                val, get_verbose_code_parts(code, guard)
+            )
+
+        add_guard("__module__", module)
+        add_guard("__qualname__", qualname)
+
+    def CONSTANT_MATCH(self, guard, metadata):
+        # self.builder.set(guard.name, metadata)
+        self.builder.CONSTANT_MATCH(guard)
+
+    def SHAPE_ENV(self, guard, metadata):
+        # TODO Share the same helper function from _dynamo/guards.py
+        from torch._dynamo.guards import _get_closure_vars
+        from torch.fx.experimental.symbolic_shapes import SYMPY_INTERP
+
+        if metadata is not None:
+            assert isinstance(metadata, list)
+            self.builder.add_python_lambda_leaf_guard_to_root(
+                metadata,
+                (),
+                closure_vars={**SYMPY_INTERP, **_get_closure_vars()},
+            )
+
+    def DETERMINISTIC_ALGORITHMS(self, guard, metadata):
+        pass
+
+    def DEFAULT_DEVICE(self, guard, metadata):
+        pass
+
+    def GRAD_MODE(self, guard, metadata):
+        pass
+
+    def TORCH_FUNCTION_STATE(self, guard, metadata):
+        pass
+
+    def SEQUENCE_LENGTH(self, guard, metadata):
+        from torch._dynamo.guards import get_verbose_code_parts
+
+        # Mostly copying from GuardBuilder.SEQUENCE_LENGTH
+        ref = self.builder.arg_ref(guard)
+        value_len, value_type = metadata
+        if not issubclass(value_type, dict):
+            # C++ DICT_LENGTH checks for type
+            self.TYPE_MATCH(guard, (value_type.__module__, value_type.__qualname__))
+
+        code = []
+        if value_len == 0:
+            code.append(f"not {ref}")
+        else:
+            code.append(f"len({ref}) == {value_len}")
+
+        if issubclass(value_type, dict):
+            self.builder.get_guard_manager(guard).add_dict_length_check_guard(
+                value_len, get_verbose_code_parts(code, guard)
+            )
+        else:
+            self.builder.get_guard_manager(guard).add_length_check_guard(
+                value_len, get_verbose_code_parts(code, guard)
+            )
+
+    def HASATTR(self, guard, metadata):
```

Comment:
what's the protocol for making sure people don't add new guards without adding this?

Would it be simpler not to have each Guard have a ser/deser method? That way, when people add new Guard types, they would have to implement serde directly, instead of finding a separate class for them.